src/solve/fulfill.rs

```diff
@@ -346,7 +346,12 @@ impl<'s> Fulfill<'s> {
     /// the outcome of type inference by updating the replacements it provides.
     pub fn solve(mut self, subst: Substitution) -> Result<Solution> {
         let outcome = self.fulfill()?;
-        if outcome.is_complete() && !self.ambiguous {
+
+        if self.cannot_prove {
```

Comment:
Instead of having a `cannot_prove` field, can we use a local variable in `fulfill()` and extend the `Outcome` enum? In that case, I think we would replace these two `ifs` with something like:

```rust
match outcome {
    Outcome::CannotProve => return Ok(Solution::CannotProve),
    Outcome::Complete => {
        // No obligations remain, so we have definitively solved our goals,
        ...
        return ...; // as before
    }
    Outcome::Incomplete => { /* fallthrough to figure out some good suggestions */ }
}
```
    }